lib/hyrax/acl/access_control_list.rb

```diff
@@ -58,12 +104,122 @@ def has_grant?(mode:, agent:)
         end
       end
 
+      ##
+      # @api public
+      #
+      # @return [Boolean]
+      def pending_changes?
+        change_set.changed?
+      end
+
       ##
       # @api public
       #
       # @return [Set<Hyrax::Permission>]
       def permissions
-        access_control.permissions
+        Set.new(change_set.permissions)
+      end
+
+      ##
+      # @api public
+      #
+      # @example
+      #    user = User.find('user_id')
+      #
+      #    acl.revoke(:read).from(user)
+      def revoke(mode)
+        ModeRevoke.new(self, mode)
+      end
+
+      ##
+      # @api public
+      #
+      # Saves the ACL for the resource, by saving each permission policy
+      #
+      # @yield [acl] gives self to the block only if the save is successful
+      # @return [Boolean]
+      def save
+        return true unless pending_changes?
+
+        change_set.sync
+        persister.save(resource: change_set.resource)
+        yield self if block_given?
+        @change_set = nil
+
+        true
+      end
+
+      private
+
+      ##
+      # @api private
+      def access_control_model
+        AccessControl.for(resource: resource, query_service: query_service)
+      end
+
+      ##
+      # @api private
+      def change_set
+        @change_set ||= AccessControlChangeSet.new(access_control_model)
+      end
+
+      ##
+      # @abstract
+      # @api private
+      class ModeEditor
+        def initialize(acl, mode)
+          @acl  = acl
+          @mode = mode.to_sym
+        end
+
+        private
+
+        ##
+        # Returns the identifier used by ACLs to identify agents.
+        #
+        # This defaults to the `:agent_key`, but if that method doesn’t exist,
+        # `:user_key` will be used as a fallback.
+        def id_for(agent:)
+          key = agent.try(:agent_key) || agent.user_key
```

Comment:
Do we need to add `user_key` attribute/method to Hyrax::Acl::Agent? It only has a single property `name`  and method `agent_key`.

Reply:
@lsitu `user_key` is for compatibility with `Hydra::User`/devise.